### DISCUSSION TITLE
Error while using verification endpoint from wallet API

### DISCUSSION THREAD

**Discussion by eminademburan** [walt-id/.github#84] [Help] [answered]
I am running the wallet, issuer and verifier API  standalone as stated in this link https://docs.oss.walt.id/wallet/api/setup/open-source. While using the verification endpoint from the wallet API,  I get an error. I am using the following endpoint.

/wallet-api/wallet/{wallet}/exchange/usePresentationRequest

You can see the error in the attached file. I do not get the error when I run the APIs using docker compose, but I get the error when I run the containers in standalone. Can you help with that?
[issue 2.txt](https://github.com/user-attachments/files/15506202/issue.2.txt)

**Comment by mikeplotean** [comment, marked as answer]
judging by the logs, _exchange/usePresentationRequest_ is provided an unresolved value for the `presentationRequest` field (the `presentation_definition` parameter is missing from the url).

Generally speaking, a presentation requires 3 steps (unlike the single step for credential claiming):
1. resolve presentation request
```
/wallet-api/wallet/{wallet}/exchange/resolvePresentationRequest
```
- the response is a complete presentation request, URL-encoded
- this url contains the resolved `presentation_definition`
2. match presentation-definition credentials
   - by providing the `presentation_definition` from step#1
```
/wallet-api/wallet/{wallet}/exchange/matchCredentialsForPresentationDefinition
```
 - the response is a list of credentials matching the presentation definition
 - however, this step can be skipped if the credential-id's are already known
3. create the presentation
   - by providing the did
   - and the resolved presentation-request-url (which should include the `presentation_definition` parameter)
   - and the list of credential-id's
   - optionally, the disclosed properties and a note can also be provided
```
/wallet-api/wallet/{wallet}/exchange/usePresentationRequest
```

You can also resolve the presentation-definition manually by fetching it from the `presentation_definition_uri` that can be found in the original presentation-request-url, attach it as a value for the `presentation_definition` parameter and supply this updated presentation-request-url to the _exchange/usePresentationRequest_ api. However, this might not work for all requests.

walt.id documentation provides more details on [how to present a verifiable credential using oid4vc](https://docs.oss.walt.id/wallet/api/credentials/guides/present-vc-oid4vp).

**Comment by eminademburan** [comment]
Thank you for your help @mikeplotean. I skipped the first step and I think it was the reason why I got this error. The issue was resolved.

**Comment by waltkb** [comment]
Good job! @mikeplotean
